frontend/dockerfile/parser/parser_heredoc_test.go

```diff
@@ -0,0 +1,275 @@
+content 1
+FILE1
+content 2
+FILE2
+
+COPY <<X <<Y /dest
+Y
+X
+X
+Y
+	`)
+
+	tests := []map[string]Heredoc{
+		nil, // ENV EXAMPLE=bla
+		nil, // RUN ls
+		nil, // USER <<INVALID
+		nil, // INVALID
+		{
+			// RUN <<EMPTY
+			"EMPTY": {
+				Name:    "EMPTY",
+				Content: "",
+				Expand:  true,
+			},
+		},
+		{
+			// RUN <<EMPTY2
+			"EMPTY2": {
+				Name:           "EMPTY2",
+				Content:        "",
+				Expand:         true,
+				FileDescriptor: 3,
+			},
+		},
+		nil, // RUN "<<NOHEREDOC"
+		{
+			// RUN <<INDENT
+			"INDENT": {
+				Name:    "INDENT",
+				Content: "\tfoo\n\tbar\n",
+				Expand:  true,
+			},
+		},
+		{
+			// RUN <<-UNINDENT
+			"UNINDENT": {
+				Name:    "UNINDENT",
+				Content: "\tbaz\n\tquux\n",
+				Expand:  true,
+				Chomp:   true,
+			},
+		},
+		{
+			// RUN <<-UNINDENT2
+			"UNINDENT2": {
+				Name:    "UNINDENT2",
+				Content: "\tbaz\n\tquux\n",
+				Expand:  true,
+				Chomp:   true,
+			},
+		},
+		{
+			// RUN <<-EXPAND
+			"EXPAND": {
+				Name:    "EXPAND",
+				Content: "\texpand $NAME\n",
+				Expand:  true,
+				Chomp:   true,
+			},
+		},
+		{
+			// RUN <<-'NOEXPAND'
+			"NOEXPAND": {
+				Name:    "NOEXPAND",
+				Content: "\tdon't expand $NAME\n",
+				Expand:  false,
+				Chomp:   true,
+			},
+		},
+		{
+			// RUN <<COPY
+			"COPY": {
+				Name:    "COPY",
+				Content: "echo hello world\necho foo bar\n",
+				Expand:  true,
+			},
+		},
+		{
+			// RUN <<COMMENT
+			"COMMENT": {
+				Name:    "COMMENT",
+				Content: "# internal comment\necho hello world\necho foo bar # trailing comment\n",
+				Expand:  true,
+			},
+		},
+		{
+			// RUN <<MOUNT
+			"MOUNT": {
+				Name:    "MOUNT",
+				Content: "echo hello\n",
+				Expand:  true,
+			},
+		},
+		{
+			// COPY <<FILE1 <<FILE2 /dest
+			"FILE1": {
+				Name:    "FILE1",
+				Content: "content 1\n",
+				Expand:  true,
+			},
+			"FILE2": {
+				Name:    "FILE2",
+				Content: "content 2\n",
+				Expand:  true,
+			},
+		},
+		{
+			// COPY <<X <<Y /dest
+			"X": {
+				Name:    "X",
+				Content: "Y\n",
+				Expand:  true,
+			},
+			"Y": {
+				Name:    "Y",
+				Content: "X\n",
+				Expand:  true,
+			},
+		},
+	}
+
+	result, err := Parse(dockerfile)
+	require.NoError(t, err)
+
+	for i, test := range tests {
+		child := result.AST.Children[i+1]
+		require.Equal(t, test, child.Heredocs)
+	}
+}
+
+func TestParseJSONHeredoc(t *testing.T) {
+	dockerfile := bytes.NewBufferString(`
+FROM alpine:3.6
+
+RUN ["whoami"]
+RUN ["<<EOF"]
+RUN ["<<'EOF'"]
+	`)
+
+	result, err := Parse(dockerfile)
+	require.NoError(t, err)
+
+	for i := 1; i <= 3; i++ {
+		child := result.AST.Children[i]
+		require.Nil(t, child.Heredocs)
+	}
+}
+
+func TestHeredocChomp(t *testing.T) {
+	heredoc := Heredoc{
+		Name:    "EOF",
+		Content: "\thello\n\tworld\n",
+	}
+	require.Equal(t, "\thello\n\tworld\n", heredoc.GetContent())
+	heredoc.Chomp = true
+	require.Equal(t, "hello\nworld\n", heredoc.GetContent())
+}
+
+func TestParseHeredocHelpers(t *testing.T) {
```

Comment:
We should (in addition) test finding the heredoc from the line. Move the `shlex.ProcessWords` to a separate function so it can be unit tested.

Reply:
Or I guess another way is to add more cases to `TestParseExtractsHeredoc`